contracts/core/src/token/OmniBridgeNative.sol

```diff
@@ -83,15 +84,19 @@ contract OmniBridgeNative is OwnableUpgradeable {
 
     /**
      * @notice Withdraw `amount` native OMNI to `to`. Only callable via xcall from OmniBridgeL1.
+     * @param payor     The address of the account with OMNI on L1.
+     * @param to        The address to receive the OMNI on Omni.
+     * @param amount    The amount of OMNI to withdraw.
+     * @param l1Balance The OMNI balance of the L1 bridge contract, synced on each withdraw.
      */
-    function withdraw(address payor, address to, uint256 amount) external {
+    function withdraw(address payor, address to, uint256 amount, uint256 l1Balance) external {
         XTypes.MsgContext memory xmsg = omni.xmsg();
 
         require(msg.sender == address(omni), "OmniBridge: not xcall"); // this protects against reentrancy
         require(xmsg.sender == l1Bridge, "OmniBridge: not bridge");
         require(xmsg.sourceChainId == l1ChainId, "OmniBridge: not L1");
 
-        l1BridgeBalance += amount;
+        l1BridgeBalance = l1Balance;
```

Comment:
This is really the whole fix (plus the corresponding change on the l1 bridge)

The rests is just tests / comments / refactoring out encodeWithSelector